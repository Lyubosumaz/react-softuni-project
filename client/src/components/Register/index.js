import { connect } from 'react-redux';
import { useState } from 'react';
import { PropTypes } from 'prop-types';
import { history } from '../../utils/history';
import { httpUser } from '../../services/http';
import schema from './register-validations';
import { setNotification } from '../Notification/actions';
import { formComponent, formFieldsWrapper, buttonClass } from '../../utils/class-names.json';
import { factoryButtons } from '../../utils/factory';

function Register({ setNotificationSuccess, setNotificationError }) {
    const username = useFormInput('');
    const email = useFormInput('');
    const password = useFormInput('');
    const repeatPassword = useFormInput('');
    const [subscribe, setSubscribe] = useState(false);
    const [errors, setErrors] = useState({});

    function useFormInput(initialValue) {
        const [value, setValue] = useState(initialValue);

        function handleChange(event) {
            setValue(event.target.value);
            validate(event);
        }

        function validate(event) {
            const name = event.target.id;

            schema.fields[name]
                .validate(event.target.value, { abortEarly: false })
                .then(() => {
                    setErrors({ ...errors, [name]: [] });
                })
                .catch((err) => {
                    setErrors({ ...errors, [name]: err.errors });
                });
        }

        return { value, onChange: handleChange };
    }

    function handleRegister(e) {
        e.preventDefault();

        const data = {
            username: username.value,
            email: email.value,
            password: password.value,
            repeatPassword: repeatPassword.value,
            subscribe: subscribe,
        };

        const hasErrors = Object.keys(errors).filter((key) => errors[key].length > 0);

        if (data.password !== data.repeatPassword) {
            return setErrors({ ...errors, repeatPassword: ["Passwords doesn't match"] });
        }

        if (hasErrors.length === 0 && data.username && data.email && data.password) {
            httpUser
                .register(data)
                .then((res) => {
                    setNotificationSuccess(res);
                    history.push('/login');
                })
                .catch((err) => {
                    setNotificationError(err);
                });
        }
    }

    return (
        <section className={`${formComponent}`}>
            <div className={`${formFieldsWrapper}`}>
                <form onSubmit={handleRegister}>
                    <div className="form-field">
                        <label htmlFor="username">
                            <span>Username:</span>
                        </label>

                        <input type="text" placeholder="Enter your Username" className="form-input" id="username" {...username} />

                        {Array.isArray(errors.username) && errors.username[0] ? (
                            <div className="form-error-message">
                                <span>{errors.username[0]}</span>
                            </div>
                        ) : (
                            <div className="form-error-container">
                                <span>Error Container</span>
                            </div>
                        )}
                    </div>

                    <div className="form-field">
                        <label htmlFor="email">
                            <span>Email:</span>
                        </label>

                        <input type="email" placeholder="Enter your Email" className="form-input" id="email" {...email} />

                        {Array.isArray(errors.email) && errors.email[0] ? (
                            <div className="form-error-message">
                                <span>{errors.email[0]}</span>
                            </div>
                        ) : (
                            <div className="form-error-container">
                                <span>Error Container</span>
                            </div>
                        )}
                    </div>

                    <div className="form-field">
                        <label htmlFor="password">
                            <span>Password:</span>
                        </label>

                        <input type="password" placeholder="Enter your Password" className="form-input" id="password" {...password} />

                        {Array.isArray(errors.password) && errors.password[0] ? (
                            <div className="form-error-message">
                                <span>{errors.password[0]}</span>
                            </div>
                        ) : (
                            <div className="form-error-container">
                                <span>Error Container</span>
                            </div>
                        )}
                    </div>

                    <div className="form-field">
                        <label htmlFor="repeat-password">
                            <span>Repeat-Password:</span>
                        </label>

                        <input type="password" placeholder="Confirm your Password" className="form-input" id="repeatPassword" {...repeatPassword} />

                        {Array.isArray(errors.repeatPassword) && errors.repeatPassword[0] ? (
                            <div className="form-error-message">
                                <span>{errors.repeatPassword[0]}</span>
                            </div>
                        ) : (
                            <div className="form-error-container">
                                <span>Error Container</span>
                            </div>
                        )}
                    </div>

                    <div className="form-field-actions">
                        <div className="terms-actions">
                            <span>By creating an account, you agree to our:</span>
                            {factoryButtons({ buttonStyles: buttonClass.Register })('terms-and-conditions', 'Terms & Conditions', 'terms')}
                        </div>

                        <div className="subscribe-actions">
                            <span>I do accept to receive feedback on my email:</span>
                            <span>
                                <input type="checkbox" id="register-checkbox" onClick={(e) => setSubscribe(e.target.checked)} />
                                <label htmlFor="register-checkbox">Subscribe</label>
                            </span>
                        </div>
                    </div>

                    <div className="form-field-buttons">
                        <input type="reset" className="form-action-btn" value="reset" />
                        <input type="submit" className="form-action-btn" value="register" />
                    </div>
                </form>
            </div>
        </section>
    );
}

function mapDispatchToProps(dispatch) {
    return {
        setNotificationSuccess: (data) => dispatch(setNotification().success(data)),
        setNotificationError: (data) => dispatch(setNotification().error(data)),
    };
}

export default connect(null, mapDispatchToProps)(Register);

Register.propTypes = {
    setNotificationSuccess: PropTypes.func.isRequired,
    setNotificationError: PropTypes.func.isRequired,
};
