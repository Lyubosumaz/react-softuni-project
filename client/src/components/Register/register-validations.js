import * as Yup from 'yup';

// Yup.addMethod(Yup.string, equalTo(ref: ))

const schema = Yup.object().shape({
    username: Yup
        .string('Username should be String')
        .required('Username is required')
        .min(4, 'Username should be at least 4 characters long'),

    email: Yup
        .string('Email should be String')
        .required('Email is required')
        .email('Please enter your email address'),

    password: Yup
        .string('Password should be String')
        .required('Password is required')
        .min(5, 'Password should be at least 5 character'),

    repeatPassword: Yup
        .string('Repeated password should be String')
        // .oneOf([Yup.ref('password', null)], 'Repeated password doesn\'t match the password')
        .required('Repeated-Password is required')
        .min(5, 'Repeated-Password should be at least 5 character'),
});

export default schema;