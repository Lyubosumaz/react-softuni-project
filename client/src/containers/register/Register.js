import React, { useState } from 'react';
import userService from '../../services/user-services';
import schema from './register-validator';


export default function Register() {
    const username = useFormInput('');
    const email = useFormInput('');
    const password = useFormInput('');
    const repeatPassword = useFormInput('');
    const [errors, setErrors] = useState({});

    return (
        <form>
            <div className="form-container">
                <h1>Register</h1>
                <p>Please fill this form to create your Account.</p>
                <div className="form-div-container">
                    <label for="username"><b>Username:</b></label>
                    <input type="text" placeholder="Enter your Username" className="form-input" id="username" {...username} />
                    {errors.username && <div className="form-input-error">{errors.username[0]}</div>}
                </div>
                <div className="form-div-container">
                    <label for="email"><b>Email:</b></label>
                    <input type="text" placeholder="Enter your Email" className="form-input" id="email" {...email} />
                    {errors.email && <div className="form-input-error">{errors.email[0]}</div>}
                </div>
                <div className="form-div-container">
                    <label for="password"><b>Password:</b></label>
                    <input type="password" placeholder="Enter your Password" className="form-input" id="password" {...password} />
                    {errors.password && <div className="form-input-error">{errors.password[0]}</div>}
                </div>
                <div className="form-div-container">
                    <label for="repeat-password"><b>Repeat-Password:</b></label>
                    <input type="password" placeholder="Confirm your Password" className="form-input" id="repeatPassword" {...repeatPassword} />
                    {errors.repeatPassword && <div className="form-input-error">{errors.repeatPassword[0]}</div>}
                </div>
                <div>
                    <p>By creating an account, you agree to our <a href="/terms-and-privacy">Terms & Conditions</a>.</p>
                    <button type="button" className="form-action-btn" onClick={handleSubmit}>Register</button>
                </div>
                <div className="form-info-container">
                    <p>Already have account? <a href="/login">Sign in</a>.</p>
                </div>
            </div>
        </form>
    );

    function handleSubmit() {
        const data = {
            username: username.value,
            email: email.value,
            password: password.value,
            repeatPassword: repeatPassword.value
        }

        const hasErrors = Object.keys(errors).filter(key => errors[key].length > 0)

        if (data.password !== data.repeatPassword) {
            console.log('Passwords doesn\'t match')
            return;
        }

        if (hasErrors.length === 0 && data.username && data.email && data.password) {
            userService.register(data).then(() => {
                console.log('Redirect')
            });
        }
    }

    function useFormInput(initialValue) {
        const [value, setValue] = useState(initialValue);

        function handleChange(ev) {
            setValue(ev.target.value);
            validate(ev)
        }

        function validate(ev) {
            const name = ev.target.id;
            schema.fields[name].validate(ev.target.value, { abortEarly: false })
                .then(() => {
                    setErrors({ ...errors, [name]: [] });
                })
                .catch((err) => {
                    setErrors({ ...errors, [name]: err.errors })
                });
        }

        return {
            value,
            onChange: handleChange
        };
    }
}



