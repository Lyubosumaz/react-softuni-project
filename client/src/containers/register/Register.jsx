import { useState } from 'react';
import history from '../../utils/history';
import http from '../../services/http';
import { toast } from 'react-toastify';
import schema from './register-validations';
// import './register.css';

export default function Register() {
    const username = useFormInput('');
    const email = useFormInput('');
    const password = useFormInput('');
    const repeatPassword = useFormInput('');
    const [subscribe, setSubscribe] = useState(false);
    const [errors, setErrors] = useState({});

    const handleRoute = (name) => (e) => {
        e.preventDefault();
        history.push(name);
    };

    function useFormInput(initialValue) {
        const [value, setValue] = useState(initialValue);

        function handleChange(event) {
            setValue(event.target.value);
            validate(event);
        }

        function validate(event) {
            const name = event.target.id;

            schema.fields[name]
                .validate(event.target.value, { abortEarly: false })
                .then(() => {
                    setErrors({ ...errors, [name]: [] });
                })
                .catch((err) => {
                    setErrors({ ...errors, [name]: err.errors });
                });
        }

        return { value, onChange: handleChange };
    }

    console.log(errors);

    function handleSubmit(e) {
        e.preventDefault();

        const data = {
            username: username.value,
            email: email.value,
            password: password.value,
            repeatPassword: repeatPassword.value,
            subscribe: subscribe,
        };

        const hasErrors = Object.keys(errors).filter((key) => errors[key].length > 0);

        if (data.password !== data.repeatPassword) {
            return setErrors({ ...errors, repeatPassword: ["Passwords doesn't match"] });
        }

        if (hasErrors.length === 0 && data.username && data.email && data.password) {
            http.User.register(data)
                .then((res) => {
                    toast(res.message, {
                        type: toast.TYPE.SUCCESS,
                    });
                    history.push('/login');
                })
                .catch((err) => {
                    toast(err.message, {
                        type: toast.TYPE.ERROR,
                    });
                });
        }
    }

    return (
        <section className="form-container">
            <header className="form-header">
                <h1>Register</h1>
                <p>Please fill this form to create your Account.</p>
            </header>

            <div className="form-fields-wrapper">
                <form onSubmit={handleSubmit}>
                    <div className="form-field">
                        <label htmlFor="username">
                            <span>Username:</span>
                        </label>

                        <input type="text" placeholder="Enter your Username" className="form-input" id="username" {...username} />

                        {Array.isArray(errors.username) && errors.username[0] ? (
                            <div className="form-error-message">
                                <span>{errors.username[0]}</span>
                            </div>
                        ) : (
                            <div className="form-error-container">
                                <span>Error Container</span>
                            </div>
                        )}
                    </div>

                    <div className="form-field">
                        <label htmlFor="email">
                            <span>Email:</span>
                        </label>

                        <input type="email" placeholder="Enter your Email" className="form-input" id="email" {...email} />

                        {Array.isArray(errors.email) && errors.email[0] ? (
                            <div className="form-error-message">
                                <span>{errors.email[0]}</span>
                            </div>
                        ) : (
                            <div className="form-error-container">
                                <span>Error Container</span>
                            </div>
                        )}
                    </div>

                    <div className="form-field">
                        <label htmlFor="password">
                            <span>Password:</span>
                        </label>

                        <input type="password" placeholder="Enter your Password" className="form-input" id="password" {...password} />

                        {Array.isArray(errors.password) && errors.password[0] ? (
                            <div className="form-error-message">
                                <span>{errors.password[0]}</span>
                            </div>
                        ) : (
                            <div className="form-error-container">
                                <span>Error Container</span>
                            </div>
                        )}
                    </div>

                    <div className="form-field">
                        <label htmlFor="repeat-password">
                            <span>Repeat-Password:</span>
                        </label>

                        <input type="password" placeholder="Confirm your Password" className="form-input" id="repeatPassword" {...repeatPassword} />

                        {Array.isArray(errors.repeatPassword) && errors.repeatPassword[0] ? (
                            <div className="form-error-message">
                                <span>{errors.repeatPassword[0]}</span>
                            </div>
                        ) : (
                            <div className="form-error-container">
                                <span>Error Container</span>
                            </div>
                        )}
                    </div>

                    <div className="form-field-actions">
                        <div className="terms">
                            <span>By creating an account, you agree to our</span>
                            <button className="info-button" onClick={handleRoute('/terms-and-conditions')}>
                                Terms &amp; Conditions
                            </button>
                        </div>

                        <div className="subscribe">
                            <span>I do accept to receive feedback on my email:</span>
                            <input type="checkbox" id="register-checkbox" onClick={(e) => setSubscribe(e.target.checked)} />
                            <label htmlFor="register-checkbox">Subscribe</label>
                        </div>
                    </div>

                    <div className="form-field-buttons">
                        <input type="reset" className="form-action-btn" value="reset" />
                        <input type="submit" className="form-action-btn" value="register" />
                    </div>
                </form>
            </div>

            <div className="info-container">
                <span>Already have account?</span>
                <button className="info-button" onClick={handleRoute('/login')}>
                    Sign in
                </button>
            </div>
        </section>
    );
}
