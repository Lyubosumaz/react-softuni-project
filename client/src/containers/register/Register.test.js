import React from 'react';
import { shallow } from 'enzyme';
import Register from './Register';

it('Renter Register component', () => {
    const wrapper = shallow(<Register />);
    expect(wrapper).toBeTruthy();
});

it('Checks for html changes', () => {
    const wrapper = shallow(<Register />);
    expect(wrapper).toMatchSnapshot();
});
