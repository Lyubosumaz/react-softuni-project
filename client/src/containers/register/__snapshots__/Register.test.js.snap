// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checks for html changes 1`] = `
<div
  className="main-container"
>
  <form>
    <h1>
      Register
    </h1>
    <p>
      Please fill this form to create your Account.
    </p>
    <div
      className="form-div-container"
    >
      <label
        htmlFor="username"
      >
        <b>
          Username:
        </b>
      </label>
      <input
        className="form-input"
        id="username"
        onChange={[Function]}
        placeholder="Enter your Username"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-div-container"
    >
      <label
        htmlFor="email"
      >
        <b>
          Email:
        </b>
      </label>
      <input
        className="form-input"
        id="email"
        onChange={[Function]}
        placeholder="Enter your Email"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-div-container"
    >
      <label
        htmlFor="password"
      >
        <b>
          Password:
        </b>
      </label>
      <input
        className="form-input"
        id="password"
        onChange={[Function]}
        placeholder="Enter your Password"
        type="password"
        value=""
      />
    </div>
    <div
      className="form-div-container"
    >
      <label
        htmlFor="repeat-password"
      >
        <b>
          Repeat-Password:
        </b>
      </label>
      <input
        className="form-input"
        id="repeatPassword"
        onChange={[Function]}
        placeholder="Confirm your Password"
        type="password"
        value=""
      />
    </div>
    <div>
      <p>
        By creating an account, you agree to our 
        <button
          className="info-button"
          onClick={[Function]}
        >
          Terms & Conditions
        </button>
        .
      </p>
      <p
        className="register-checkbox"
      >
        I do accept to receive feedback on my email:
        <input
          onClick={[Function]}
          type="checkbox"
        />
        <b>
          Subscribe
        </b>
        .
      </p>
    </div>
    <div>
      <button
        className="form-action-btn"
        onClick={[Function]}
        type="button"
      >
        Register
      </button>
    </div>
    <div
      className="info-container"
    >
      <p>
        Already have account? 
        <button
          className="info-button"
          onClick={[Function]}
        >
          Sign in
        </button>
        .
      </p>
    </div>
  </form>
</div>
`;
